### PR TITLE
Automatically reset postgres connection when closed

### LINKED ISSUE #560: Refresh postgres connection if closed
If any of the low-level tokio-postgres methods fail due to a [closed connection](https://docs.rs/tokio-postgres/latest/tokio_postgres/error/struct.Error.html#method.is_closed), we should automatically replace the connection in the `SqlStorage` instance with a fresh one. The connection can be closed for a number of reasons, including the server restarting.

### PR DESCRIPTION
**Pull request by jbearer**
Closes #560

### CODE REVIEW

**Comment by imabdulbasit** [comment]
Looks good! this just needs to be rebased to  include some new changes related to TmpDb

File: src/data_source/storage/sql/client.rs [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+            .unwrap();
+        assert_eq!(modified, 2);
+        assert_eq!(get_values().await, [1, 2, 3, 4]);
+
+        // Shut down the database.
+        db.stop();
+
+        // Subsequent statements and queries within the transaction fail.
+        tracing::info!("query with DB shut down");
+        let err = client
```
Comment (imabdulbasit) [inline src/data_source/storage/sql/client.rs:474]: if this transaction fails then shouldn't the `in_progress` be set to `false`. After this, if I begin a new transaction then that should create a new connection or return an error if that times out?

Comment (imabdulbasit) [reply src/data_source/storage/sql/client.rs:474]: or maybe if I create a new transaction and the connection is closed then it retries ?

Comment (jbearer) [reply src/data_source/storage/sql/client.rs:474]: Yes, but we haven't started a new transaction yet. We try that at the end of this test and make sure statements in the new transaction succeed

**Review by imabdulbasit — APPROVED** [review]
LGTM

**Pull request merged**